### PR TITLE
Multithreading fixes and improvements

### PR DESCRIPTION
**Pull request by xvallspl**
Fixing some spotted bugs/inefficiencies in TMVA and the mathematical libraries.

### CODE REVIEW

**Comment by dpiparo** [comment]
@phsft-bot build please

**Comment by ashlaban** [comment]
## Benchmarking of changes

Platform:
Intel(R) Core(TM) i7-4578U CPU @ 3.00GHz

`189ca14cbdbc262a910cf7d479ae11f6f0082f84` (previous "random" checkout):

`root -l -b -q 'TMVAClassification.C("BDTG")'  3.94s user 0.78s system 149% cpu 3.152 total`
`root -l -b -q 'TMVARegression.C("BDTG")'  12.70s user 2.15s system 148% cpu 10.012 total`

`bd26ea689b4c1cecba7e12ad101d906185b6d90d` (after xavi's PR merged):

`root -l -b -q 'TMVAClassification.C("BDTG")'  3.65s user 0.87s system 145% cpu 3.102 total`
`root -l -b -q 'TMVARegression.C("BDTG")'  12.89s user 2.22s system 146% cpu 10.335 total`

## Output comparison

Top line is "old" commit, bottom line is "new" commit.

```
DataSet       MVA
Name:         Method:          ROC-integ
dataset       BDTG           : 0.897
dataset       BDTG           : 0.897

DataSet              MVA              Signal efficiency: from test sample (from training sample)
Name:                Method:          @B=0.01             @B=0.10            @B=0.30
dataset              BDTG           : 0.206 (0.540)       0.705 (0.858)      0.902 (0.928)
dataset              BDTG           : 0.206 (0.540)       0.705 (0.858)      0.902 (0.928)

Evaluation results ranked by smallest RMS on test sample:
dataset              BDTG           :    0.156    0.159     2.51     1.92  |  3.118  3.186
dataset              BDTG           :    0.198    0.165     2.60     2.00  |  3.084  3.149
```

In summary, the changes do not to change the output significantly. There is a slight change for regression output. @lmoneta, can you judge the relevance of this?

For time consumption, no significant changes observed.
